Haskell/Haskell.sublime-syntax

```diff
@@ -1,407 +1,1388 @@
-        |scanl|scanl1|scanr|scanr1|seq|sequence|sequenceA|sequence_|show|showChar|showList
-        |showParen|showString|shows|showsPrec|significand|signum|sin|sinh|snd|span|splitAt|sqrt
-        |subtract|succ|sum|tail
-        |take|takeWhile|tan|tanh|toEnum|toInteger|toRational|traverse|truncate|uncurry
-        |undefined|unlines|until|unwords|unzip|unzip3|userError|words
-        |writeFile
-        |zip|zip3|zipWith|zipWith3
-        ) \b
-      scope: support.function.prelude.haskell
-    - match: '[[:lower:]][^\s{{operator_char}}),\]{}]+'
-      scope: meta.name.haskell
-
-  comment:
-    # As of build 4079 the [:punct:] class is missing some ASCII chars that
-    # Unicode considers part of the Symbol category.
-    - match: '--(?![[:punct:]!-/:-@\[-`{-~])'
-      scope: punctuation.definition.comment.haskell
-      push:
-        - meta_scope: comment.line.double-dash.haskell
-        - match: $\n?
-          pop: true
-    - include: block_comment
+  statement-terminators:
+    # Depending on layout, semicolon may be needed to terminate statements.
+    # https://www.haskell.org/onlinereport/haskell2010/haskellch2.html#x7-210002.7
+    - match: ';'
+      scope: punctuation.terminator.statement.haskell
 
-  block_comment:
-    - match: '\{-(?!#)'
-      scope: punctuation.definition.comment.begin.haskell
-      push:
-        - meta_scope: comment.block.haskell
-        - match: '\{-#'
-          push:
-            - match: '-\}'
-              pop: true
-            - include: block_comment
-        - include: block_comment
-        - match: '-\}'
-          scope: punctuation.definition.comment.end.haskell
-          pop: true
-
-  module_exports:
-    - match: \(
-      scope: punctuation.section.group.begin.haskell
-      push:
-        - meta_scope: meta.declaration.exports.haskell
-        - match: \)
-          scope: punctuation.section.group.end.haskell
-          pop: true
-        - match: '\b[a-z][a-zA-Z_''0-9]*'
-          scope: variable.function.haskell
-        - match: '\b[A-Z][A-Za-z_''0-9]*'
-          scope: storage.type.haskell
-        - match: ','
-          scope: punctuation.separator.comma.haskell
-        - include: operator_paren
-        - match: \(.*?\)
-          comment: So named because I don't know what to call this.
-          scope: meta.other.unknown.haskell
-  
-  module_name:
-    - match: '\b{{module_name}}\b'
-      scope: support.other.module.haskell
-  
-  type_signature:
-    - include: pragma
-    - match: '(?:->|→)'
-      scope: keyword.other.arrow.haskell
-    - match: '(?:=>|⇒)'
-      scope: keyword.other.big-arrow.haskell
-    - match: '\b[a-z][a-zA-Z0-9_'']*\b'
-      scope: variable.other.generic-type.haskell
-    - match: '\b[A-Z][a-zA-Z0-9_'']*\b'
-      scope: storage.type.haskell
-    - match: \(\)
-      scope: support.constant.unit.haskell
+  sequence-separators:
+    - match: ',|\|(?!\|)'
+      scope: punctuation.separator.sequence.haskell
```

Comment:
The syntax doesn't push into those contexts to be able to distinguish that. It was just too error prone.

Reply:
I don't think a single pipe character is valid anywhere else. Commas we can easily defer.

Reply:
They are separators in mappings if I don't miss something. That's one major reason for the scope decision.

https://github.com/sublimehq/Packages/blob/01624ba13f8fe351184f119a70c49836da2a7130/Haskell/tests/syntax_test_haskell.hs#L1295-L1344

Otherwise they are separators of a list of guarded expressions

https://github.com/sublimehq/Packages/blob/01624ba13f8fe351184f119a70c49836da2a7130/Haskell/tests/syntax_test_haskell.hs#L2068-L2085

Erlang also has that concept of guarded clauses/expressions which I use `punctuation.separator` for, too, in order to scope the separator which is `;` in that case. Erlang even calls it a list of guarded clauses. Hence my impression is that this concept works for Haskell, too.

At least it helps to distinguish normal operators used in guard expressions from guard boundaries a bit better.

Reply:
Also

```Haskell
    -- Declare a list-like instance for Char
    data instance XList Char = XCons !Char !(XList Char) | XNil
```

Reply:
Right, they are used for sum types …
Let's track these instances that require context-awareness in (a) separate issue(s).

Reply:
Also mentioned this in https://github.com/sublimehq/Packages/issues/2868. You may resolve the discussion if you only kept it open for tracking this.